### PR TITLE
#17 Import QR code from photos

### PR DESCRIPTION
**Pull request by benbusby**
Addresses #17, adding the ability to import a QR code from the user's photo album.

### CODE REVIEW

File: Tofu/AccountsViewController.swift [hunk 1]
```diff
@@ -76,12 +80,19 @@ class AccountsViewController: UITableViewController {
...
 
         let scanQRCode = UIAlertAction(title: "Scan QR Code", style: .default) { [unowned self] _ in
             self.performSegue(withIdentifier: "ScanSegue", sender: self)
         }
+        
+        let importQRCode = UIAlertAction(title: "Import QR Image", style: .default) { [unowned self] _ in
+            self.qrImporter.allowsEditing = false
+            self.qrImporter.sourceType = .photoLibrary
+                
+            self.present(self.qrImporter, animated: true, completion: nil)
```
Comment (calleerlandsson) [inline Tofu/AccountsViewController.swift]: Did you consider calling [`isSourceTypeAvailable(_:)`](https://developer.apple.com/documentation/uikit/uiimagepickercontroller/1619144-issourcetypeavailable) to check if we can pick from the photo library before presenting this controller as suggested by the [docs on UIImagePickerController](https://developer.apple.com/documentation/uikit/uiimagepickercontroller)? I'm not sure what happens if we present it anyway and the docs don't seem to mention this.

File: Tofu/AccountsViewController.swift [hunk 2]
```diff
@@ -123,6 +135,26 @@ class AccountsViewController: UITableViewController {
...
+    func imagePickerController(_ picker: UIImagePickerController,
+                               didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
+        guard let selectedQRCode = info[UIImagePickerController.InfoKey.originalImage] as? UIImage,
+            let detector = CIDetector(ofType: CIDetectorTypeQRCode,
+                                      context: nil,
+                                      options: [CIDetectorAccuracy: CIDetectorAccuracyHigh]),
+            let ciImage = CIImage(image: selectedQRCode),
+            let features = detector.features(in: ciImage) as? [CIQRCodeFeature] else { return }
+
+        let qrCodeURL = URL(string: features.reduce("") { $0 + ($1.messageString ?? "") })
```
Comment (calleerlandsson) [inline Tofu/AccountsViewController.swift]: Did you ever see more then one feature returned? We seem to be able to provide a [CIDetectorMaxFeatureCount](https://developer.apple.com/documentation/coreimage/cidetectormaxfeaturecount) option to specify how many we want at most. The docs do however say this defaults to 1.

If so, what do you think about including this in the guard statement? Maybe something like this:

```
guard ...,
    let features = detector.features(in: ciImage) as? [CIQRCodeFeature],
    let messageString = features.first?.messageString,
    let qrCodeURL = URL(string: messageString) else { return }

File: Tofu/AccountsViewController.swift [hunk 3]
```diff
@@ -123,6 +135,26 @@ class AccountsViewController: UITableViewController {
...
+
+        if qrCodeURL != nil, let account = Account(url: qrCodeURL!) {
```
Comment (calleerlandsson) [inline Tofu/AccountsViewController.swift]: Maybe `let account = Account(url: qrCodeURL)` could go in the `guard` too?

File: Tofu/AccountsViewController.swift [hunk 4]
```diff
@@ -123,6 +135,26 @@ class AccountsViewController: UITableViewController {
             navigationItem.leftBarButtonItem = editButtonItem
         }
     }
+    
+    // MARK: UIImagePickerControlDelegate
+    
+    func imagePickerController(_ picker: UIImagePickerController,
+                               didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
+        guard let selectedQRCode = info[UIImagePickerController.InfoKey.originalImage] as? UIImage,
```
Comment (calleerlandsson) [inline Tofu/AccountsViewController.swift]: If this guard fails, should we present a message to the user? Something like "Failed to Detect QR Code"?

File: Tofu/AccountsViewController.swift (hunk 1 above)
Comment (benbusby) [reply Tofu/AccountsViewController.swift]: Ah, no I hadn't. I hadn't actually considered instances where it wouldn't be available -- that's good to know for future reference though. Just added it in.

File: Tofu/AccountsViewController.swift (hunk 2 above)
Comment (benbusby) [reply Tofu/AccountsViewController.swift]: I did not, but I agree that including it in the guard is a good idea.

**Comment by benbusby** [comment]
All questions/comments should be addressed in my latest commit. Thank you for reviewing! I haven't written any Swift in a while, so I appreciate your pointers. Let me know if there's anything else that needs addressing.

**Comment by calleerlandsson** [comment]
Great work @benbusby! Thanks again!

**Pull request merged**